File: training-data/typescript/pub_sub.ts
type Handler<T> = (message: T) => void;

class PubSub<T = unknown> {
  private subscribers = new Map<string, Handler<T>[]>();

  subscribe(topic: string, handler: Handler<T>): void {
    if (!this.subscribers.has(topic)) this.subscribers.set(topic, []);
    this.subscribers.get(topic)!.push(handler);
  }

  publish(topic: string, message: T): void {
    for (const handler of this.subscribers.get(topic) ?? []) handler(message);
  }
}

const bus = new PubSub<string>();
bus.subscribe('news', (msg) => console.log(`received: ${msg}`));
bus.publish('news', 'hello subscribers');

File: training-data/rust/pub_sub.rs
use std::collections::HashMap;

struct PubSub {
    subscribers: HashMap<String, Vec<Box<dyn Fn(&str)>>>,
}

impl PubSub {
    fn new() -> Self {
        PubSub { subscribers: HashMap::new() }
    }

    fn subscribe(&mut self, topic: &str, handler: Box<dyn Fn(&str)>) {
        self.subscribers.entry(topic.to_string()).or_insert_with(Vec::new).push(handler);
    }

    fn publish(&self, topic: &str, message: &str) {
        if let Some(handlers) = self.subscribers.get(topic) {
            for handler in handlers {
                handler(message);
            }
        }
    }
}

fn main() {
    let mut bus = PubSub::new();
    bus.subscribe("news", Box::new(|msg| println!("received: {}", msg)));
    bus.publish("news", "hello subscribers");
}
